libcore/IconInfo.vala

```diff
@@ -0,0 +1,339 @@
+/* Copyright (c) 2018 elementary LLC (https://elementary.io)
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, Inc.,; either version 2 of
+ * the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+public class Marlin.IconInfo : GLib.Object {
+    // TODO: Replace me with real scale factor and adapt the widgets
+    private const int SCALE = 1;
+
+    private int64 last_use_time;
+    private Gdk.Pixbuf? pixbuf;
+    private string icon_name;
+
+    public Marlin.IconInfo.for_pixbuf (Gdk.Pixbuf? pixbuf) {
+        this.pixbuf = pixbuf;
+    }
+
+    public Marlin.IconInfo.for_icon_info (Gtk.IconInfo icon_info) {
+        try {
+            pixbuf = icon_info.load_icon ();
+        } catch (Error e) {
+            critical (e.message);
+        }
+
+        var filename = icon_info.get_filename ();
+        if (filename != null) {
+            filename = GLib.Path.get_basename (filename);
+            var last = filename.last_index_of_char ('.');
+            icon_name = filename.substring (0, last);
+        }
+    }
+
+    construct {
+        last_use_time = GLib.get_monotonic_time ();
+        schedule_reap_cache ();
```

Comment:
Not sure why?

Reply:
On reflection, this replaces the reaping that in the c version was triggered by pixbuf_toggle_notify (i.e when a pixbuf became unused for a certain time).